demos/fleet_provisioning/fleet_provisioning_with_csr/pkcs11_operations.c

```diff
@@ -0,0 +1,1162 @@
+    certificateTemplate.xCertificateType.pValue = &certificateType;
+    certificateTemplate.xCertificateType.ulValueLen = sizeof( CK_CERTIFICATE_TYPE );
+    certificateTemplate.xTokenObject.type = CKA_TOKEN;
+    certificateTemplate.xTokenObject.pValue = &tokenStorage;
+    certificateTemplate.xTokenObject.ulValueLen = sizeof( tokenStorage );
+
+    result = C_GetFunctionList( &functionList );
+
+    /* Test for a valid certificate: 0x2d is '-', as in ----- BEGIN CERTIFICATE. */
+    if( ( certificate == NULL ) || ( certificate[ 0 ] != 0x2d ) )
+    {
+        result = CKR_ATTRIBUTE_VALUE_INVALID;
+    }
+
+    if( result == CKR_OK )
+    {
+        /* Convert the certificate to DER format if it was in PEM. The DER key
+         * should be about 3/4 the size of the PEM key, so mallocing the PEM key
+         * size is sufficient. */
+        derObject = ( uint8_t * ) malloc( certificateTemplate.xValue.ulValueLen );
+        derLen = certificateTemplate.xValue.ulValueLen;
+
+        if( derObject != NULL )
+        {
+            conversion = convert_pem_to_der( ( unsigned char * ) certificateTemplate.xValue.pValue,
+                                             certificateTemplate.xValue.ulValueLen,
+                                             derObject, &derLen );
+
+            if( 0 != conversion )
+            {
+                result = CKR_ARGUMENTS_BAD;
+            }
+        }
+        else
+        {
+            result = CKR_HOST_MEMORY;
+        }
+    }
+
+    if( result == CKR_OK )
+    {
+        /* Set the template pointers to refer to the DER converted objects. */
+        certificateTemplate.xValue.pValue = derObject;
+        certificateTemplate.xValue.ulValueLen = derLen;
+    }
+
+    /* Best effort clean-up of the existing object, if it exists. */
+    if( result == CKR_OK )
+    {
+        destroyProvidedObjects( session, &label, &certificateClass, 1 );
+    }
+
+    /* Create an object using the encoded client certificate. */
+    if( result == CKR_OK )
+    {
+        LogInfo( ( "Writing certificate..." ) );
+
+        result = functionList->C_CreateObject( session,
+                                               ( CK_ATTRIBUTE_PTR ) &certificateTemplate,
+                                               sizeof( certificateTemplate ) / sizeof( CK_ATTRIBUTE ),
+                                               &objectHandle );
+    }
+
+    if( derObject != NULL )
+    {
+        free( derObject );
+    }
+
+    return result;
+}
+
+/*-----------------------------------------------------------*/
+
+bool loadClaimCredentials( CK_SESSION_HANDLE p11Session )
+{
+    bool status;
+    char claimCert[ CLAIM_CERT_BUFFER_LENGTH ] = { 0 };
+    size_t claimCertLength = 0;
+    char claimPrivateKey[ CLAIM_PRIVATE_KEY_BUFFER_LENGTH ] = { 0 };
+    size_t claimPrivateKeyLength = 0;
+    CK_RV ret;
+
+    status = readFile( CLAIM_CERT_PATH, claimCert, CLAIM_CERT_BUFFER_LENGTH,
```

Comment:
This information should also be taken as input parameters of the claim certificate file path